enable dmvr AVX2 code

nuomi2021 opened the pull request.

nuomi2021 committed:
add github workflow

nuomi2021 committed:
avcodec/vvcdec: Use av_image_copy_plane for DMVR 10-bit integer pixels

It's no need to shift and interpolate for 10-bit integer pixels,
av_image_copy_plane is enough

nuomi2021 committed:
x86/vvcdec: add dmvr avx2 code

QSXW commented on libavcodec/vvc/inter_template.c:
```diff
@@ -483,6 +484,10 @@ static void FUNC(apply_bdof)(uint8_t *_dst, const ptrdiff_t _dst_stride, int16_t
     const int height, const intptr_t mx, const intptr_t my, const int width)
 {
+#if BIT_DEPTH == 10
+    av_image_copy_plane((uint8_t*)dst, sizeof(int16_t) * MAX_PB_SIZE, _src, _src_stride,
```
Do you think it will look better if we move it to the end of the function so we can distinguish it from the variable declaration?

QSXW commented on libavcodec/x86/vvc/vvc_dmvr.asm:
```diff
@@ -0,0 +1,398 @@
+    pmullw                 %6, %2, %4
+    paddw                  %1, %6
+%if bd == 12
+    psrlw                  %1, 1                        ;avoid signed mul for pmulhrsw
```
Better to add a space between `;` and `comments`

QSXW commented on tests/checkasm/vvc_mc.c:
```diff
@@ -324,6 +324,64 @@ static void check_avg(void)
+                            case 0: type = "dmvr"; break; // 0 0
+                            case 1: type = "dmvr_h"; break; // 0 1
+                            case 2: type = "dmvr_v"; break; // 1 0
+                            case 3: type = "dmvr_hv"; break; // 1 1
```
Maybe align the `break;` and `comments`?

QSXW commented:
Others LGTM

stone-d-chen commented on libavcodec/x86/vvc/vvc_dmvr.asm:
```diff
@@ -0,0 +1,396 @@
+    mov                   wd, wm
+
+%if %1 > 10
+    vpbroadcastw          m4, [shift_%1]
```
I think it's generally preferred to do

` vpbroadcastd          m4, [shift_%1]`

and then `shift_12: times 2 dw 1 << (15 - (12 - 10))`

nuomi2021 replied:
Thank you for the review.
Any benefit for this. Alignment?

stone-d-chen replied:
On intel architectures it's faster especially for Alderlake it's about 3x throughput. broadcastw also uses at least two execution ports so it's possible that you get more performance using broadcastd depending on what other ports instructions near it need. For Zen 3/4 performance is identical.

Alder Lake P
broadcastd
Latency = 5/8
Throughput = 0.33

broadcastw
Lat = 6/9
Throughput = 1

stone-d-chen replied:
![image](https://github.com/user-attachments/assets/bb922bcd-4a3f-490f-8a6d-f6e34046ddc2)

QSXW replied:
Yeah. `vpbroadcastd` will have better latency than  `vpbroadcastw`.

nuomi2021 replied:
done

nuomi2021 replied:
@stone-d-chen the page is cute, where can I find it?

stone-d-chen replied:
https://uops.info/table.html

It's great!

nuomi2021 committed:
checkasm: add tests for vvc dmvr

dmvr_8_12x20_c: 186.2
dmvr_8_12x20_avx2: 25.7
dmvr_8_20x12_c: 181.7
dmvr_8_20x12_avx2: 25.2
dmvr_8_20x20_c: 283.2
dmvr_8_20x20_avx2: 32.0
dmvr_10_12x20_c: 90.0
dmvr_10_12x20_avx2: 15.7
dmvr_10_20x12_c: 41.0
dmvr_10_20x12_avx2: 14.7
dmvr_10_20x20_c: 81.5
dmvr_10_20x20_avx2: 26.7
dmvr_12_12x20_c: 190.7
dmvr_12_12x20_avx2: 20.2
dmvr_12_20x12_c: 187.2
dmvr_12_20x12_avx2: 20.2
dmvr_12_20x20_c: 292.7
dmvr_12_20x20_avx2: 27.2
dmvr_h_8_12x20_c: 317.0
dmvr_h_8_12x20_avx2: 37.0
dmvr_h_8_20x12_c: 340.0
dmvr_h_8_20x12_avx2: 41.0
dmvr_h_8_20x20_c: 540.7
dmvr_h_8_20x20_avx2: 64.0
dmvr_h_10_12x20_c: 322.7
dmvr_h_10_12x20_avx2: 30.7
dmvr_h_10_20x12_c: 344.2
dmvr_h_10_20x12_avx2: 34.0
dmvr_h_10_20x20_c: 529.0
dmvr_h_10_20x20_avx2: 51.5
dmvr_h_12_12x20_c: 326.7
dmvr_h_12_12x20_avx2: 33.5
dmvr_h_12_20x12_c: 331.7
dmvr_h_12_20x12_avx2: 51.2
dmvr_h_12_20x20_c: 534.0
dmvr_h_12_20x20_avx2: 62.7
dmvr_hv_8_12x20_c: 650.0
dmvr_hv_8_12x20_avx2: 57.2
dmvr_hv_8_20x12_c: 676.2
dmvr_hv_8_20x12_avx2: 70.0
dmvr_hv_8_20x20_c: 1068.5
dmvr_hv_8_20x20_avx2: 103.2
dmvr_hv_10_12x20_c: 649.0
dmvr_hv_10_12x20_avx2: 48.2
dmvr_hv_10_20x12_c: 677.7
dmvr_hv_10_20x12_avx2: 59.7
dmvr_hv_10_20x20_c: 1093.5
dmvr_hv_10_20x20_avx2: 91.7
dmvr_hv_12_12x20_c: 660.0
dmvr_hv_12_12x20_avx2: 58.7
dmvr_hv_12_20x12_c: 682.7
dmvr_hv_12_20x12_avx2: 72.0
dmvr_hv_12_20x20_c: 1094.0
dmvr_hv_12_20x20_avx2: 113.2
dmvr_v_8_12x20_c: 325.7
dmvr_v_8_12x20_avx2: 31.2
dmvr_v_8_20x12_c: 326.2
dmvr_v_8_20x12_avx2: 38.5
dmvr_v_8_20x20_c: 538.5
dmvr_v_8_20x20_avx2: 54.2
dmvr_v_10_12x20_c: 318.5
dmvr_v_10_12x20_avx2: 23.7
dmvr_v_10_20x12_c: 330.7
dmvr_v_10_20x12_avx2: 40.5
dmvr_v_10_20x20_c: 567.5
dmvr_v_10_20x20_avx2: 48.0
dmvr_v_12_12x20_c: 335.2
dmvr_v_12_12x20_avx2: 30.0
dmvr_v_12_20x12_c: 330.2
dmvr_v_12_20x12_avx2: 39.5
dmvr_v_12_20x20_c: 535.2
dmvr_v_12_20x20_avx2: 60.0

nuomi2021 commented:
sent to upstream for review https://patchwork.ffmpeg.org/project/ffmpeg/patch/TYSPR06MB64333950CE2429A5119F3F36AAAB2@TYSPR06MB6433.apcprd06.prod.outlook.com/

nuomi2021 commented:
fix dmvr for https://github.com/ffvvc/FFmpeg/issues/51

nuomi2021 commented:
merged as https://github.com/FFmpeg/FFmpeg/commit/7eb1df44ae6daea1fa68ff2fdd96a13d3111be3e
thank you for the review
